File: src/platform/android/app/src/main/java/com/nearclip/services/ble/ProtocolBufferMessageHandler.kt
package com.nearclip.services.ble

import android.util.Log
import com.google.protobuf.InvalidProtocolBufferException
import nearclip.protocol.BLEMessage
import nearclip.protocol.ConnectionState
import nearclip.protocol.DeviceInfo
import nearclip.protocol.DeviceType
import nearclip.protocol.MessageType
import java.nio.ByteBuffer

/**
 * Protocol Buffers消息处理器
 * 统一Android和Mac端的消息序列化/反序列化实现
 */
class ProtocolBufferMessageHandler {

    companion object {
        private const val TAG = "ProtocolBufferHandler"
    }

    /**
     * 序列化消息为字节数组
     */
    fun serializeMessage(message: TestMessage): ByteArray {
        return try {
            val protoMessage = BLEMessage.newBuilder()
                .setMessageId(message.messageId)
                .setType(convertToProtoMessageType(message.type))
                .setPayload(message.payload)
                .setTimestamp(message.timestamp)
                .setSequenceNumber(message.sequenceNumber)
                .build()

            protoMessage.toByteArray()
        } catch (e: Exception) {
            Log.e(TAG, "序列化消息失败: ${e.message}", e)
            // 回退到旧的字符串序列化方式
            fallbackSerializeMessage(message)
        }
    }

    /**
     * 反序列化字节数组为消息
     */
    fun deserializeMessage(data: ByteArray): TestMessage? {
        return try {
            val protoMessage = BLEMessage.parseFrom(data)
            TestMessage(
                messageId = protoMessage.messageId,
                type = convertFromProtoMessageType(protoMessage.type),
                payload = protoMessage.payload,
                timestamp = protoMessage.timestamp,
                sequenceNumber = protoMessage.sequenceNumber
            )
        } catch (e: InvalidProtocolBufferException) {
            Log.w(TAG, "Protocol Buffers解析失败，尝试回退方式: ${e.message}")
            // 回退到旧的字符串反序列化方式
            fallbackDeserializeMessage(data)
        } catch (e: Exception) {
            Log.e(TAG, "反序列化消息失败: ${e.message}", e)
            null
        }
    }

    /**
     * 转换消息类型为Protocol Buffers枚举
     */
    private fun convertToProtoMessageType(type: MessageType): MessageTypeProto {
        return when (type) {
            MessageType.PING -> MessageTypeProto.MESSAGE_TYPE_PING
            MessageType.PONG -> MessageTypeProto.MESSAGE_TYPE_PONG
            MessageType.DATA -> MessageTypeProto.MESSAGE_TYPE_DATA
            MessageType.ACK -> MessageTypeProto.MESSAGE_TYPE_ACK
        }
    }

    /**
     * 从Protocol Buffers枚举转换消息类型
     */
    private fun convertFromProtoMessageType(type: MessageTypeProto): MessageType {
        return when (type) {
            MessageTypeProto.MESSAGE_TYPE_PING -> MessageType.PING
            MessageTypeProto.MESSAGE_TYPE_PONG -> MessageType.PONG
            MessageTypeProto.MESSAGE_TYPE_DATA -> MessageType.DATA
            MessageTypeProto.MESSAGE_TYPE_ACK -> MessageType.ACK
            MessageTypeProto.MESSAGE_TYPE_UNSPECIFIED -> MessageType.DATA // 默认值
            else -> MessageType.DATA // 未知类型默认为DATA
        }
    }

    /**
     * 回退序列化方式（向后兼容）
     */
    private fun fallbackSerializeMessage(message: TestMessage): ByteArray {
        val data = "${message.messageId}|${message.type}|${message.payload}|${message.timestamp}|${message.sequenceNumber}"
        return data.toByteArray(Charsets.UTF_8)
    }

    /**
     * 回退反序列化方式（向后兼容）
     */
    private fun fallbackDeserializeMessage(data: ByteArray): TestMessage? {
        return try {
            val messageString = String(data, Charsets.UTF_8)
            val parts = messageString.split("|")

            if (parts.size >= 5) {
                TestMessage(
                    messageId = parts[0],
                    type = MessageType.valueOf(parts[1]),
                    payload = parts[2],
                    timestamp = parts[3].toLongOrNull() ?: System.currentTimeMillis(),
                    sequenceNumber = parts[4].toIntOrNull() ?: 0
                )
            } else {
                null
            }
        } catch (e: Exception) {
            Log.e(TAG, "回退反序列化也失败: ${e.message}", e)
            null
        }
    }

    /**
     * 验证消息格式
     */
    fun validateMessage(message: TestMessage): Boolean {
        return try {
            // 基本字段验证
            message.messageId.isNotBlank() &&
            message.type != null &&
            message.timestamp > 0 &&
            message.sequenceNumber >= 0 &&
            message.payload.length <= 1024 // 限制载荷大小
        } catch (e: Exception) {
            Log.e(TAG, "消息验证失败: ${e.message}", e)
            false
        }
    }

    /**
     * 创建设备信息Proto消息
     */
    fun createDeviceInfoProto(device: BleDevice): ByteArray {
        return try {
            val protoDevice = DeviceInfo.newBuilder()
                .setDeviceId(device.deviceId)
                .setDeviceName(device.deviceName)
                .setDeviceType(if (device.deviceName.contains("Android", ignoreCase = true))
                    DeviceType.DEVICE_TYPE_ANDROID else DeviceType.DEVICE_TYPE_MAC)
                .setRssi(device.rssi)
                .setIsNearclipDevice(device.deviceName.contains("NearClip", ignoreCase = true))
                .build()

            protoDevice.toByteArray()
        } catch (e: Exception) {
            Log.e(TAG, "创建设备信息Proto失败: ${e.message}", e)
            ByteArray(0)
        }
    }
}

// 为了避免命名冲突，重命名Protocol Buffers生成的枚举
typealias MessageTypeProto = nearclip.protocol.MessageType

File: src/platform/mac/NearClip/Sources/Services/ProtocolBufferMessageHandler.swift
import Foundation
import CoreBluetooth

/**
 * Protocol Buffers消息处理器
 * 统一Mac和Android端的消息序列化/反序列化实现
 */
class ProtocolBufferMessageHandler {

    // MARK: - Message Serialization

    /// 序列化消息为Data
    func serializeMessage(_ message: TestMessage) -> Data {
        do {
            var protoMessage = BLEMessage()
            protoMessage.messageID = message.messageId
            protoMessage.type = convertToProtoMessageType(message.type)
            protoMessage.payload = message.payload
            protoMessage.timestamp = message.timestamp
            protoMessage.sequenceNumber = Int32(message.sequenceNumber)

            return try protoMessage.serializedData()
        } catch {
            print("❌ Protocol Buffers序列化失败: \(error.localizedDescription)")
            // 回退到旧的字符串序列化方式
            return fallbackSerializeMessage(message)
        }
    }

    /// 反序列化Data为消息
    func deserializeMessage(_ data: Data) -> TestMessage? {
        do {
            let protoMessage = try BLEMessage(serializedData: data)
            return TestMessage(
                messageId: protoMessage.messageID,
                type: convertFromProtoMessageType(protoMessage.type),
                payload: protoMessage.payload,
                timestamp: protoMessage.timestamp,
                sequenceNumber: Int(protoMessage.sequenceNumber)
            )
        } catch {
            print("⚠️ Protocol Buffers解析失败，尝试回退方式: \(error.localizedDescription)")
            // 回退到旧的字符串反序列化方式
            return fallbackDeserializeMessage(data)
        }
    }

    // MARK: - Type Conversion

    /// 转换消息类型为Protocol Buffers枚举
    private func convertToProtoMessageType(_ type: MessageType) -> MessageTypeProto {
        switch type {
        case .ping:
            return .ping
        case .pong:
            return .pong
        case .data:
            return .data
        case .ack:
            return .ack
        }
    }

    /// 从Protocol Buffers枚举转换消息类型
    private func convertFromProtoMessageType(_ type: MessageTypeProto) -> MessageType {
        switch type {
        case .ping:
            return .ping
        case .pong:
            return .pong
        case .data:
            return .data
        case .ack:
            return .ack
        case .unspecified:
            return .data // 默认值
        case .UNRECOGNIZED:
            return .data // 未知类型默认为DATA
        }
    }

    // MARK: - Fallback Methods (向后兼容)

    /// 回退序列化方式（向后兼容）
    private func fallbackSerializeMessage(_ message: TestMessage) -> Data {
        let messageString = "\(message.messageId)|\(message.type.rawValue)|\(message.payload)|\(message.timestamp)|\(message.sequenceNumber)"
        return messageString.data(using: .utf8) ?? Data()
    }

    /// 回退反序列化方式（向后兼容）
    private func fallbackDeserializeMessage(_ data: Data) -> TestMessage? {
        guard let messageString = String(data: data, encoding: .utf8) else { return nil }

        let parts = messageString.split(separator: "|")
        guard parts.count >= 5 else { return nil }

        return TestMessage(
            messageId: String(parts[0]),
            type: MessageType(rawValue: String(parts[1])) ?? .data,
            payload: String(parts[2]),
            timestamp: Double(String(parts[3])) ?? Date().timeIntervalSince1970,
            sequenceNumber: Int(String(parts[4])) ?? 0
        )
    }

    // MARK: - Validation

    /// 验证消息格式
    func validateMessage(_ message: TestMessage) -> Bool {
        // 基本字段验证
        return !message.messageId.isEmpty &&
               message.timestamp > 0 &&
               message.sequenceNumber >= 0 &&
               message.payload.count <= 1024 // 限制载荷大小
    }

    /// 创建设备信息Proto消息
    func createDeviceInfoProto(from peripheral: CBPeripheral, rssi: NSNumber) -> Data {
        do {
            var protoDevice = DeviceInfo()
            protoDevice.deviceID = peripheral.identifier.uuidString
            protoDevice.deviceName = peripheral.name ?? "未知设备"
            protoDevice.rssi = Int32(truncating: rssi)
            protoDevice.isNearclipDevice = peripheral.name?.contains("NearClip") ?? false

            // 根据设备名称判断设备类型
            if peripheral.name?.contains("Android", options: .caseInsensitive) ?? false {
                protoDevice.deviceType = .android
            } else if peripheral.name?.contains("Mac", options: .caseInsensitive) ?? false ||
                      peripheral.name?.contains("iPhone", options: .caseInsensitive) ?? false {
                protoDevice.deviceType = .mac
            } else {
                protoDevice.deviceType = .other
            }

            return try protoDevice.serializedData()
        } catch {
            print("❌ 创建设备信息Proto失败: \(error.localizedDescription)")
            return Data()
        }
    }

    // MARK: - Connection State Conversion

    /// 转换连接状态为Protocol Buffers枚举
    func convertToProtoConnectionState(_ state: ConnectionState) -> ConnectionStateProto {
        switch state {
        case .disconnected:
            return .disconnected
        case .connecting:
            return .connecting
        case .connected:
            return .connected
        case .failed:
            return .failed
        }
    }

    /// 从Protocol Buffers枚举转换连接状态
    func convertFromProtoConnectionState(_ state: ConnectionStateProto) -> ConnectionState {
        switch state {
        case .disconnected:
            return .disconnected
        case .connecting:
            return .connecting
        case .connected:
            return .connected
        case .failed:
            return .failed
        case .unspecified:
            return .disconnected // 默认值
        case .UNRECOGNIZED:
            return .disconnected // 未知状态默认为断开
        }
    }
}

// 为了避免命名冲突，使用类型别名
typealias MessageTypeProto = Nearclip_Protocol_MessageType
typealias ConnectionStateProto = Nearclip_Protocol_ConnectionState